enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,891 @@
+
+Controllers in Cluster API respect the pause mechanism to prevent them from taking any action on the resource.
+This matches the requirements for the non-authoritative resources described in this document.
+
+The migration controller will leverage the pause mechanism to ensure that the Cluster API controllers do not take any action on the resource when it is non-authoritative.
+Admission control will ensure that the value cannot be changed by users when the resource is non-authoritative.
+
+#### Filtering reconciles in Machine API controllers
+
+Each controller will, at the start of the reconcile logic, check for the presence of the `status.authoritativeAPI` and will either skip or continue with the reconciliation based on the presence of this field.
+
+The controller will set the `Paused` condition to `True` when the resource is non-authoritative and will not perform any further reconciliation.
+
+#### The Cluster Autoscaler in mixed MachineSet clusters
+
+The Cluster Autoscaler today looks at Machine API MachineSets and filters these based on a number of annotations to identify node groups that it may scale.
+In the future, where we have both Machine API and Cluster API authoritative MachineSets, the Autoscaler must be updated to ensure it scales the authoritative MachineSet.
+
+Since every Machine API MachineSet will be reflected into Cluster API, the autoscaler can be updated to reconcile only the Cluster API MachineSets.
+When a scaling decision is made, it must then update the authoritative MachineSet to the correct scale.
+
+To do this, there are two possible approaches to consider:
+- Look up the Machine API MachineSet mirror to determine the authoritative API and update accordingly
+  - This would be a carry patch to the autoscaler, which, may cause maintenance toil over time
+- Allow the Autoscaler to update the Cluster API MachineSet, but allow the sync controller to identify the update has come from the Autoscaler, and mirror it to the authoritative MachineSet as appropriate
+  - This could be achieved by observing the managed fields property of the replicas field and determine which actor updated it.
+  If the Autoscaler last changed the replica count, it should be mirrored back to the Machine API MachineSet if it is authoritative.
+  - This creates an exception to the rules outline above and adds complication to the conversion, but, may be less of a burden than carrying a patch in the autoscaler
+
+Both avenues above will need to be explored further before a decision is made on how the Autoscaler will handle mixed MachineSet clusters.
+
+### Risks and Mitigations
+
+#### Synchronisation logic will need to be solid before it gets into the hand of customers
+
+To ensure that we do not break customer clusters, we need to make sure that our logic that converts the resources between Machine API and Cluster API is thoroughly tested.
```

Comment:
Since we're allowing two different controllers to handle the overall migration process, it's possible we could have the conversion controller shipping and converting things before enabling the authoritative checks, right?

That way we could have telemetry for those customers who have opted in to see what their conversions are looking like, whether they're overall healthy or there are glaring problems that we need to address/add to the test suite.